File: examples/haitun-workspace/skills/fusion-flow-next/src/core-ir.ts
/** Catalog-owned concept referenced by workflow constants and operators. */
export class Concept {
  public constructor(readonly name: string) {}
}

/** A typed identity or literal declared or used by a workflow. */
export class Constant {
  public constructor(
    readonly symbol: string,
    readonly belongConcepts: readonly Concept[],
  ) {}
}

/** Catalog-owned operator signature. */
export class Operator {
  public constructor(
    readonly name: string,
    readonly inputConcepts: readonly Concept[] = [],
    readonly outputConcept: Concept | null = null,
  ) {}

  public get arity(): number {
    return this.inputConcepts.length;
  }
}

/** An operator applied to recursive term arguments. */
export class CompoundTerm {
  readonly arguments: readonly Term[];

  public constructor(
    readonly operator: Operator,
    arguments_: readonly Term[],
  ) {
    this.arguments = arguments_;
  }
}

/** Ordered list values remain ordinary terms in Workflow Core IR. */
export class ListTerm {
  public constructor(readonly items: readonly Term[]) {}
}

export type RelationSymbol = "=" | "!=" | "<" | "<=" | ">" | ">=";

/** Atomic relation between two terms. */
export class Assertion {
  public constructor(
    readonly lhs: Term,
    readonly rhs: Term,
    readonly relationSymbol: RelationSymbol = "=",
  ) {}
}

export type LogicalConnective = "NOT" | "AND" | "OR";

/** Workflow condition built from assertions with NOT, AND, or OR. */
export class ConnectiveFormula {
  readonly formulaLeft: Formula;
  readonly connective: LogicalConnective;
  readonly formulaRight: Formula | null;

  public constructor(
    formulaLeft: Formula,
    connective: "NOT",
    formulaRight?: null,
  );
  public constructor(
    formulaLeft: Formula,
    connective: "AND" | "OR",
    formulaRight: Formula,
  );
  public constructor(
    formulaLeft: Formula,
    connective: LogicalConnective,
    formulaRight: Formula | null = null,
  ) {
    if (connective === "NOT" && formulaRight !== null) {
      throw new TypeError("NOT cannot have a right formula");
    }
    if (connective !== "NOT" && formulaRight === null) {
      throw new TypeError(`${connective} requires a right formula`);
    }

    this.formulaLeft = formulaLeft;
    this.connective = connective;
    this.formulaRight = formulaRight;
  }
}

export type Term = Constant | CompoundTerm | ListTerm;
export type Formula = Assertion | ConnectiveFormula;

/** Named workflow block passed between parser and checker. */
export class Workflow {
  public constructor(
    readonly name: string,
    readonly assertions: readonly Assertion[],
  ) {}
}

File: examples/haitun-workspace/skills/fusion-flow-next/fusion_flow_next/core_ir.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Literal


@dataclass(frozen=True, slots=True)
class Concept:
    name: str


@dataclass(frozen=True, slots=True)
class Constant:
    symbol: str
    belong_concepts: tuple[Concept, ...] = ()


@dataclass(frozen=True, slots=True)
class Operator:
    name: str
    input_concepts: tuple[Concept, ...] = ()
    output_concept: Concept | None = None

    @property
    def arity(self) -> int:
        return len(self.input_concepts)


@dataclass(frozen=True, slots=True)
class CompoundTerm:
    operator: Operator
    arguments: tuple[Term, ...]


@dataclass(frozen=True, slots=True)
class ListTerm:
    items: tuple[Term, ...]


@dataclass(frozen=True, slots=True)
class Assertion:
    lhs: Term
    rhs: Term
    relation_symbol: RelationSymbol = "="


@dataclass(frozen=True, slots=True)
class ConnectiveFormula:
    formula_left: Formula
    connective: LogicalConnective
    formula_right: Formula | None = None

    def __post_init__(self) -> None:
        if self.connective == "NOT" and self.formula_right is not None:
            raise ValueError("NOT cannot have a right formula")
        if self.connective != "NOT" and self.formula_right is None:
            raise ValueError(f"{self.connective} requires a right formula")


@dataclass(frozen=True, slots=True)
class IfTerm:
    condition: Formula
    when_true: Term
    when_false: Term


@dataclass(frozen=True, slots=True)
class Workflow:
    name: str
    assertions: tuple[Assertion, ...]


@dataclass(frozen=True, slots=True)
class WorkflowFile:
    constants: tuple[Constant, ...]
    workflows: tuple[Workflow, ...]


type RelationSymbol = Literal["=", "!=", "<", "<=", ">", ">="]
type LogicalConnective = Literal["NOT", "AND", "OR"]
type Term = Constant | CompoundTerm | ListTerm | IfTerm
type Formula = Assertion | ConnectiveFormula
